File: day7.py

def getCoords(input, maxX):
    xOrds = [0] * (maxX + 1)
    for x in input:
        xOrds[x]+= 1
    return xOrds

def solveForX(xOrds, targetX):
    totalMoves = 0
    for x in range(len(xOrds)):
        count = xOrds[x]
        d = abs(x - targetX)
        fuelBurned = (d * (d + 1))/2
        totalMoves+= fuelBurned * count
    return totalMoves

def getMinMoves(xOrds):
    minMoves = float('inf')
    for i in range(len(xOrds)):
        moves = solveForX(xOrds, i)
        if moves < minMoves:
            minMoves = moves
    return minMoves

def main():
    f = open("day7.txt","r")
    lines = f.readlines()
    lines = lines[0].split(',')
    lines = list(map(lambda x: int(x), lines))
    maxX = 0
    for i in range(len(lines)):
        if lines[i] > maxX:
            maxX = lines[i]
    
    xOrds = getCoords(lines, maxX)

    minMoves = getMinMoves(xOrds)
    
    
    print(minMoves)
    return

main()

File: day7.js
const fs = require('fs');

const getFuel = (xOrds, targetX) => {
    let fuel = 0;
    for (const [x, count] of Object.entries(xOrds)) {
        const d = Math.abs(x - targetX);
        const singleFuel = (d * (d+1)) / 2;
        if (targetX == 5) {
            console.log(singleFuel)
        }
        fuel = fuel + singleFuel * count
    }
    return fuel;
}

const getMinFuel = () => {
    let inputs = fs.readFileSync('day7Test.txt', 'utf8').split(',');
    let maxX = 0;
    inputs = inputs.map(x => {
        newX = parseInt(x);
        if (x > maxX) {
            maxX = x;
        }
        return parseInt(x);
    });
    const xOrds = new Array(maxX).fill(0);
    for (let i = 0; i < inputs.length; i++) {
        xOrds[inputs[i]]++;
    }
    console.log(maxX);


    let minFuel = Number.MAX_SAFE_INTEGER
    for (const [x, count] of Object.entries(xOrds)) {
        let fuel = getFuel(xOrds, x);
        if (fuel < minFuel) {
            minFuel = fuel;
        }
    }
    console.log(minFuel);
}

getMinFuel()